venafi/provider.go

```diff
@@ -94,106 +156,123 @@ Example:
 
 func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
 
-	tflog.Info(ctx, "Configuring provider\n")
-	apiKey := d.Get("api_key").(string)
-	url := d.Get("url").(string)
-	tppUser := d.Get("tpp_username").(string)
-	tppPassword := d.Get("tpp_password").(string)
-	accessToken := d.Get("access_token").(string)
-	zone := d.Get("zone").(string)
-	tflog.Info(ctx, fmt.Sprintf("====ZONE==== : %s", zone))
-	devMode := d.Get("dev_mode").(bool)
-	trustBundle := d.Get("trust_bundle").(string)
+	tflog.Info(ctx, "Configuring venafi provider")
+	apiKey := d.Get(providerApiKey).(string)
+	url := d.Get(providerURL).(string)
+	tppUser := d.Get(providerUsername).(string)
+	tppPassword := d.Get(providerPassword).(string)
+	accessToken := d.Get(providerAccessToken).(string)
+	zone := d.Get(providerZone).(string)
+	trustBundle := d.Get(providerTrustBundle).(string)
+	p12Certificate := d.Get(providerP12Cert).(string)
+	p12Password := d.Get(providerP12Password).(string)
+	clientID := d.Get(providerClientID).(string)
+
+	// Normalize zone for VCert usage
+	zone = normalizeZone(zone)
+
+	//Dev Mode
+	devMode := d.Get(providerDevMode).(bool)
+	// TLSPDC auth methods
+	userPassMethod := tppUser != "" && tppPassword != ""
+	clientCertMethod := p12Certificate != "" && p12Password != ""
+	accessTokenMethod := accessToken != ""
+	// TLSPC auth methods
+	apiKeyMethod := apiKey != ""
 
 	// Warning or errors can be collected in a slice type
 	var diags diag.Diagnostics
 
-	var cfg vcert.Config
+	if !accessTokenMethod && !clientCertMethod && !userPassMethod && !apiKeyMethod && !devMode {
+		tflog.Error(ctx, messageVenafiNoAuthProvided)
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  messageVenafiClientInitFailed,
+			Detail:   fmt.Sprintf("%s: %s", messageVenafiConfigFailed, messageVenafiNoAuthProvided),
+		})
+		return nil, diags
+	}
 
-	zone = normalizeZone(zone)
+	cfg := vcert.Config{
+		BaseUrl:    url,
+		Zone:       zone,
+		LogVerbose: true,
+		Credentials: &endpoint.Authentication{
+			ClientId: clientID,
```

Comment:
This client ID is set for each of the if/else choices below, whereas in previous logic it was not set. Is this acceptable?

Reply:
Indeed, somehow, the provider lacked two things:
1. A default clientID, which means all tokens requested by the provider were issued under `vcert-sdk`  (default of vcert golang SDK) clientID 
2. A way to pass your clientID, which means there is no way to tell TPP who is requesting tokens, as all tokens requested by the provider go with `vcert-sdk` as client

I added ability to pass your own ID as well as set a default ID for the provider